protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,965 @@
+                        };
+
+                        break;
+                    }
+                    PeersIterState::Finished => break,
+                }
+            }
+        }
+
+        state.unwrap_or(PeersIterState::Finished)
+    }
+
+    /// Only finish the [`ClosestPeersIter`]s that yielded a peer within
+    /// `peers`.
+    ///
+    /// `peers` representing the peers that either returned or put a record.
+    ///
+    /// See [`crate::query::Query::may_finish`] for details.
+    pub fn may_finish<I>(&mut self, peers: I)
+    where
+        I: IntoIterator<Item = PeerId>
+    {
+        for peer in peers {
+            if let Some(PeerState{ initiated_by, .. }) = self.contacted_peers.get_mut(&peer) {
+                self.iters[*initiated_by].finish();
+            }
+        }
+    }
+
+    /// Immediately transitions the iterator to [`PeersIterState::Finished`].
+    pub fn finish(&mut self) {
+        for iter in &mut self.iters {
+            iter.finish();
+        }
+    }
+
+    /// Checks whether the iterator has finished.
+    pub fn is_finished(&self) -> bool {
+        for iter in &self.iters {
+            if iter.is_finished() {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Note: In the case of no adversarial peers or connectivity issues along
+    ///       any path, all paths return the same result, deduplicated through
+    ///       the `ResultIter`, thus overall `into_result` returns
+    ///       `num_results`. In the case of adversarial peers or connectivity
+    ///       issues `ClosestDisjointPeersIter` tries to return the
+    ///       `num_results` closest benign peers, but as it can not
+    ///       differentiate benign from faulty paths it as well returns faulty
+    ///       peers and thus overall returns more than `num_results` peers.
+    pub fn into_result(self) -> impl Iterator<Item = PeerId> {
+        let result_per_path= self.iters.into_iter()
+            .map(|iter| iter.into_result().map(Key::from));
+
+        ResultIter::new(self.target, result_per_path).map(Key::into_preimage)
+    }
+}
+
+/// Index into the [`ClosestDisjointPeersIter`] `iters` vector.
+#[derive(Copy, Clone, Debug)]
+struct IteratorIndex(usize);
+
+impl From<usize> for IteratorIndex {
+    fn from(i: usize) -> Self {
+        IteratorIndex(i)
+    }
+}
+
+impl From<IteratorIndex> for usize {
+    fn from(i: IteratorIndex) -> Self {
+        i.0
+    }
+}
+
+impl Add<usize> for IteratorIndex {
+    type Output = Self;
+
+    fn add(self, rhs: usize) -> Self::Output {
+        (self.0 + rhs).into()
+    }
+}
+
+impl Index<IteratorIndex> for Vec<ClosestPeersIter> {
+    type Output = ClosestPeersIter;
+
+    fn index(&self, index: IteratorIndex) -> &Self::Output {
+        &self[index.0]
+    }
+}
+
+impl IndexMut<IteratorIndex> for Vec<ClosestPeersIter> {
+    fn index_mut(&mut self, index: IteratorIndex) -> &mut Self::Output {
+        &mut self[index.0]
+    }
+}
```

Comment:
These impls don't really seem worthwhile, since `IteratorIndex` is just a a private newtype. Do they offer significant code simplifications?

Reply:
I don't think they give us any code simplification but make the indexing less error prone.

With the above one can use `IteratorIndex` to directly index into a collection of iterators, thus one does not need to do `iterators[index.0]` and can thereby not confuse it with any other defined newtype pattern (e.g. as in `iterators[not_an_index_into_iterators.0]`).

I find the complexity reasonably low and not to bloat the actual business logic and thus find it helpful. If you think differently I can change it to `iterators[index.0]`.

Reply:
> With the above one can use IteratorIndex to directly index into a collection of iterators, thus one does not need to do iterators[index.0] and can thereby not confuse it with any other defined newtype pattern (e.g. as in iterators[not_an_index_into_iterators.0]).

I would agree if this newtype were not confined to this module and there were not just a single such index (e.g. there is no `SomeOtherIndex`), but here we have a use that is both confined to a single module and a single such index being tracked. The newtype and all the trait impls thus seem a bit overkill to me but I won't object if you want to keep them in, I just wanted to get clearer on your motivation, which in principle I fully understand.